changes/db.js

```diff
@@ -13,28 +13,60 @@ function saveChanges(client, changes, next) {
     var timestamp = changes.timestamp;
     Object.keys(changes.users).forEach(function (key) {
         if (changes.users[key].changesets.length > 0) {
-            q.defer(saveChange, client, timestamp, parseInt(key, 10), changes.users[key]);
+            q.defer(saveChange, client, timestamp, parseInt(key, 10), changes.users[key], changes.replicationId);
         }
     });
     q.awaitAll(function() {
         next();
     });
 }
-
-function saveChange(client, timestamp, uid, user, callback) {
-    var insertQuery = 'INSERT INTO stats (change_at, uid, nodes, ways, relations, changesets, tags_created, tags_modified, tags_deleted) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)';
-    var params = [timestamp, uid, user.nodes, user.ways, user.relations, user.changesets, user.tags_created, user.tags_modified, user.tags_deleted];
-    dbUsers.saveUser(client, uid, user.username, function (err) {
+function saveChange(client, timestamp, uid, user, replicationId, callback) {
+    dbUsers.saveUser(client, uid, user.username, function(err) {
         if (err) {
             console.log('error inserting user object', err);
             return callback(err);
         }
-        client.query(insertQuery, params, function(err) {
+
+        var md5 = helpers.getHash(replicationId.toString() + uid.toString());
+        var checkQuery = 'SELECT id from stats WHERE id=$1';
+
+        client.query(checkQuery, [md5], function(err, result) {
             if (err) {
-                console.log('error inserting change object', err);
-                return callback(err);
+                console.log('error checking existing stats row', err);
+            }
+
+            var params = [
+                md5,
+                timestamp,
+                uid,
+                user.nodes,
+                user.ways,
+                user.relations,
+                user.changesets,
+                user.tags_created,
+                user.tags_modified,
+                user.tags_deleted
+            ];
+
+            if (result.rows.length === 0) {
+                var insertQuery = 'INSERT INTO stats (id, change_at, uid, nodes, ways, relations, changesets, tags_created, tags_modified, tags_deleted) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)';
+                client.query(insertQuery, params, function(err) {
+                    if (err) {
+                        console.log('error inserting change object', err);
+                        return callback(err);
+                    }
+                    callback();
+                });
+            } else {
+                var updateQuery = 'UPDATE stats SET change_at=$2, uid=$3, nodes=$4, ways=$5, relations=$6, changesets=$7, tags_created=$8, tags_modified=$9, tags_deleted=$10 WHERE id=$1';
```

Comment:
not sure we need to update `change_at` and `uid` - I mean it would be something terribly weird if we had the same hash for a different change_at or uid .. I guess this is okay, but not sure if we should catch this somewhere, or not update these.